core/src/main/java/io/parsingdata/metal/expression/value/reference/Ref.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.expression.value.reference;
+
+import static io.parsingdata.metal.SafeTrampoline.complete;
+import static io.parsingdata.metal.SafeTrampoline.intermediate;
+import static io.parsingdata.metal.Util.checkNotNull;
+import static io.parsingdata.metal.data.selection.ByPredicate.getAllValues;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import io.parsingdata.metal.SafeTrampoline;
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseGraph;
+import io.parsingdata.metal.data.ParseValue;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.expression.value.ValueExpression;
+import io.parsingdata.metal.token.Token;
+
+/**
+ * A {@link ValueExpression} that represents all {@link Value}s in the parse
+ * state that match a provided object. This class only has a private
+ * constructor and instead must be instantiated through one of its factory
+ * methods: {@link #nameRef} (to match on name) and {@link #tokenRef} (to
+ * match on definition).
+ * @param <T> The type of reference to match on.
+ */
+public class Ref<T> implements ValueExpression {
+
+    public final T reference;
+    public final Predicate<ParseValue> predicate;
+
+    private Ref(final T reference, final Predicate<ParseValue> predicate) {
+        this.reference = checkNotNull(reference, "reference");
+        this.predicate = checkNotNull(predicate, "predicate");
```

Comment:
My OCD is very satisfied by the fact that these two words have the same amount of characters. :relieved: